### ISSUE TITLE
UTF-16 output to UTF-8 console when compiling with GCC 10.x

### ISSUE THREAD

**Issue by Markus74a** [rhash/RHash#138]
OS: Windows 10 20H2

When compiling latest RHASH with GCC 10.2.0 (from MSYS2), the compiled executable will show console output correctly when running on MSYS2 BASH or MINTTY consoles, but when trying to run RHASH in Windows CMD, it will only ever output the first character of anything printed to screen.

The `Ā` used below is the `U+0100` character. The `tee` program used below is **not** UTF-8/UTF-16 compatible.

MSYS2 BASH:
```
# ./rhash -B --sha1
RHash v1.4.0 benchmarking...
SHA1 512 MiB calculated in 0,278 sec, 1844,582 MBps
SHA1 512 MiB calculated in 0,277 sec, 1846,886 MBps
SHA1 512 MiB calculated in 0,279 sec, 1836,765 MBps
SHA1 512 MiB calculated in 0,279 sec, 1834,511 MBps
SHA1 2048 MiB total in 1,113 sec, 1840,671 MBps, CPB=1,56

# ./rhash --sha1 *.md
2750f9850af082c1eec1dd93f74da6fce2544b76  INSTALL.md
a9993e364706816aba3e25717850c26c9cd0d89d  RĀ.md
cb79a2fe53e5642b92378975a56a138934d0b5cb  README.md
```

Windows CMD:
```
> rhash -B --sha1
RSSSSS,

>rhash --sha1 *.md
a R
```
However, if the output is redirected to a pipeline, it will show correctly:
```
> rhash --sha1 *.md | tee con
2750f9850af082c1eec1dd93f74da6fce2544b76  INSTALL.md
cb79a2fe53e5642b92378975a56a138934d0b5cb  README.md
a9993e364706816aba3e25717850c26c9cd0d89d  R─Ç.md
2750f9850af082c1eec1dd93f74da6fce2544b76  INSTALL.md
cb79a2fe53e5642b92378975a56a138934d0b5cb  README.md
a9993e364706816aba3e25717850c26c9cd0d89d  R─Ç.md

> rhash --sha1 *.md > somefile
// the file "somefile" will contain everything correctly

> rhash --sha1 *.md > con
a R
// redirecting to console this way will not work
```

This only happens when using **GCC 10.2.0** to compile RHASH. When using the older **GCC 9.2.0**, the created executable will run fine everywhere.

My assumption is that the strings are passed as UTF-16, but the console expects UTF-8, so it will treat the "0" byte following the first byte as the trailing "\0" of a string instead of the 2nd byte of the first UTF-16 character. This would truncate every output after 1 character.

**Comment by Markus74a** [comment]
I did some more testing:

**GCC 10.1.0** (from http://winlibs.com/) will produce an executable file which gives the correct output, while **GCC 10.2.0** (from http://winlibs.com/) will just output only the first char the same way the MSYS2 version does.

Running a *cross-compile* on Ubuntu Linux 20.4 (**GCC 9.3.0**) will create a working executable, while running a *cross-compile* on Arch Linux (**GCC 10.2.0**) will truncate every output after 1 char.

**Comment by rhash** [comment]
Thanks for the investigation. I reproduced the bug with gcc 10.1.0, and win_utils.o indeed is now linked to __mingw_vfwprintf, __mingw_vsnprintf, __mingw_vfprintf.

I suspect __mingw_vfwprintf works somehow differently from MSVCRT vfwprintf.

**Comment by rhash** [comment]
Commit cc28211f7e0dfff82b31d8301b094f2a159b90b4 fixes the bug by defining __USE_MINGW_ANSI_STDIO macro to 0, which switches vfwprintf() back to MSVCRT implementation.

MSVCRT vfwprintf() processes "%s" and "%S" format specifiers differently from the MinGW __mingw_vfwprintf().
GCC and MinGW *wprintf() implementations follows ANSI C99 and POSIX standards:

> **%s** - If no **l** (ell) qualifier is present, the application shall ensure that the argument is a **pointer to a character array** containing a character sequence beginning in the initial shift state. Characters from the array shall be converted as if by repeated calls to the mbrtowc() function, with the conversion state described by an mbstate_t object initialized to zero before the first character is converted, and written up to (but not including) the terminating null wide character. If the precision is specified, no more than that many wide characters shall be written. If the precision is not specified, or is greater than the size of the array, the application shall ensure that the array contains a null wide character.
If an **l** (ell) qualifier is present, the application shall ensure that the argument is a pointer to an **array of type wchar_t**. Wide characters from the array shall be written up to (but not including) a terminating null wide character. If no precision is specified, or is greater than the size of the array, the application shall ensure that the array contains a null wide character. If a precision is specified, no more than that many wide characters shall be written.
**%S** (POSIX) - Equivalent to **%ls**.

Сonversely MSVCRT historically interpret **%s** and **%S** in reverse order:
> Character and string arguments that are specified by using **c** and **s** are interpreted as char and char* by **printf** family functions, or as wchar_t and wchar_t* by **wprintf** family functions. Character and string arguments that are specified by using **C** and **S** are interpreted as wchar_t and wchar_t* by **printf** family functions, or as char and char* by **wprintf** family functions. This behavior is Microsoft-specific.

**Comment by rhash** [comment]
The fix is included into RHash v1.4.1.
